File: crates/detectord/crates/mcp_detector_daemon/src/protocol.rs
//! The daemon↔UI wire protocol: newline-delimited JSON over a Unix socket.
//!
//! The connecting OS user is derived from the socket's peer credentials
//! (`getpeereid`), never from the message — so a request is always scoped to
//! the uid the kernel reports, which the UI cannot spoof.

use serde::{Deserialize, Serialize};

use edison_detectord::ServerConfig;

fn default_true() -> bool {
    true
}

/// UI → daemon requests.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "op", rename_all = "snake_case")]
pub enum Request {
    /// Enroll this user (online handshake: validate key, cache policy + known set).
    Enroll {
        url: String,
        key: String,
        #[serde(default)]
        mcp_url: Option<String>,
        /// `None` keeps the previous selection; `Some` replaces it.
        #[serde(default)]
        agents: Option<Vec<String>>,
        /// `None` keeps the previous secret.
        #[serde(default)]
        secret: Option<String>,
        /// Apply the edison-watch install + hooks (default). Set false for a
        /// detect-only enrollment (caller does its own install).
        #[serde(default = "default_true")]
        install: bool,
        /// Arm automatic quarantine enforcement. `None` keeps the prior state;
        /// the UI sets `true` once onboarding completes so the daemon stays
        /// detect-only while the user is still reviewing during setup.
        #[serde(default)]
        armed: Option<bool>,
    },
    /// Enrollment + cached policy. `refresh` re-fetches the policy first.
    Status {
        #[serde(default)]
        refresh: bool,
    },
    /// Which agents (host apps) are present.
    ListAgents,
    /// Discovered servers, classified.
    ListServers,
    /// Dispose of a discovered server.
    Disposition {
        name: String,
        #[serde(default)]
        agent: Option<String>,
        choice: Choice,
        /// For SendToEw: submit under this name instead (rename-on-conflict).
        #[serde(default)]
        rename: Option<String>,
        /// For SendToEw: submit THIS config verbatim instead of the
        /// discovered/stored one. It is the client's authoritative
        /// credential-review result (already redacted per the user's choices, and
        /// NOT auto-templatized again — so a value the user left unmarked is sent
        /// as-is). `None` = auto-templatize the discovered/stored config. The
        /// locally retained raw config (for secret injection) is unaffected.
        #[serde(default)]
        submit_config: Option<ServerConfig>,
        /// For SendToEw: register directly (`Some(true)`) or leave the request
        /// pending approval (`Some(false)`), overriding the role-derived
        /// default. The UI offers an admin both, so their explicit choice has
        /// to survive the trip.
        #[serde(default)]
        register: Option<bool>,
    },
    /// Install the `edison-watch` entry + hooks for these agents (additive: the
    /// agents join the enrolled selection). The daemon is the only component
    /// that writes agent configs.
    ApplyIntegrations { agents: Vec<String> },
    /// Remove the `edison-watch` entry for these agents.
    RevertIntegrations { agents: Vec<String> },
    /// The text of an agent's user-scope config file, for display.
    ReadConfig { agent: String },
    /// Put quarantined servers back: one by name/fingerprint, or all of them.
    RestoreQuarantined {
        #[serde(default)]
        name: Option<String>,
    },
    /// Record a submit the CALLER performed against the backend, so the daemon's
    /// seen-store stays the single source of truth for what's known.
    MarkSeen {
        name: String,
        #[serde(default)]
        agent: Option<String>,
        /// `registered` | `requested` | `dismissed`.
        status: String,
    },
    /// Remove a discovered server from its local config, leaving seen-state
    /// alone. For callers that already submitted the server and marked it seen
    /// themselves and only need the local entry gone; `disposition` is the op
    /// that submits *and* records an outcome. Removal goes through the same
    /// writer, so Claude Code project scope, Cursor plugin dirs and the state
    /// DBs are all handled, and the entry stays restorable.
    RemoveLocal {
        name: String,
        #[serde(default)]
        agent: Option<String>,
    },
    /// Force a policy + known-set refresh.
    RefreshPolicy,
    /// Verify an existing secret key; adopt (install) it if valid.
    VerifySecret { key: String },
    /// Destructively reset to a new secret key (deletes encrypted personal
    /// values), then install it. `confirm` must be true.
    ResetSecret {
        key: String,
        #[serde(default)]
        confirm: bool,
    },
    /// Remove this user's enrollment.
    Unenroll,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Choice {
    /// Submit to Edison Watch (register/request) + remove locally.
    SendToEw,
    /// Leave quarantined; don't re-prompt.
    Skip,
}

/// daemon → UI replies (a direct answer to a [`Request`]).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "reply", rename_all = "snake_case")]
pub enum Reply {
    Status(Status),
    // Struct-shaped (not `Vec` newtypes): an internally-tagged enum can't hold a
    // bare sequence.
    Agents {
        agents: Vec<AgentInfo>,
    },
    Servers {
        servers: Vec<ServerView>,
    },
    Secret(SecretOutcome),
    Integrations {
        changes: Vec<IntegrationChange>,
    },
    Config {
        path: String,
        content: Option<String>,
    },
    Restored {
        restored: u32,
        errors: Vec<String>,
    },
    Ack,
    Error {
        message: String,
    },
}

/// The outcome of installing or removing the `edison-watch` entry for one agent.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct IntegrationChange {
    pub agent: String,
    /// The config file written, when there was one (Claude Code goes through
    /// its CLI, which owns the path).
    pub path: Option<String>,
    /// The backup taken before the first edit of that file, if any.
    pub backup_path: Option<String>,
    pub ok: bool,
    #[serde(default)]
    pub error: Option<String>,
}

/// Outcome of a verify (`valid`/`expired` set) or reset (`deleted` set).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SecretOutcome {
    #[serde(default)]
    pub valid: Option<bool>,
    #[serde(default)]
    pub expired: Option<bool>,
    #[serde(default)]
    pub deleted: Option<u32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Status {
    pub user: String,
    pub enrolled: bool,
    pub org_id: Option<String>,
    pub org_name: Option<String>,
    pub email: Option<String>,
    pub role: Option<String>,
    pub quarantine: bool,
    pub quarantined_count: usize,
    /// Whether automatic quarantine enforcement is armed (onboarding complete).
    #[serde(default)]
    pub armed: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentInfo {
    pub name: String,
    pub installed: bool,
    /// Hook bindings this agent has, and how many are already injected. The
    /// counts come from the same presence checks the injector uses, so the UI
    /// never has to open an agent's hook file to report coverage.
    #[serde(default)]
    pub hooks_total: u32,
    #[serde(default)]
    pub hooks_installed: u32,
    /// The URL of the installed `edison-watch` entry, or `None` when the agent
    /// has no entry. The UI compares it with the URL it expects instead of
    /// reading and parsing the agent's config itself.
    #[serde(default)]
    pub edison_url: Option<String>,
    /// The agent's user-scope config file, so the UI can name it (and ask for
    /// its contents via `read_config`) without resolving paths of its own.
    #[serde(default)]
    pub config_path: Option<String>,
    /// Workspace-level hook targets this agent has (e.g. one `.vscode/tasks.json`
    /// per enumerated VSCode workspace), and how many already carry the Edison
    /// Watch task. The UI renders hook coverage from these instead of walking
    /// the user's project directories itself. Zero for agents with no
    /// workspace hook surface.
    #[serde(default)]
    pub workspace_hooks_total: u32,
    #[serde(default)]
    pub workspace_hooks_installed: u32,
}

/// One discovered server instance (not deduped — carries its source path).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ServerView {
    pub name: String,
    pub agent: String,
    /// `stdio` | `http` | `opaque`.
    pub kind: String,
    /// `edison` | `known` | `new` | `opaque` | `report`.
    pub state: String,
    pub fingerprint: Option<String>,
    pub path: String,
    /// The server's launch config, so a UI (e.g. onboarding) can render and act
    /// on it without re-discovering locally. `None` on older/omitted views.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub config: Option<ServerConfig>,
}

/// daemon → UI unsolicited pushes. (Delivery is wired with the supervisor;
/// defined here so the contract is fixed.)
#[allow(dead_code)]
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "event", rename_all = "snake_case")]
pub enum Event {
    Quarantined(ServerView),
    Discovered(ServerView),
    PolicyChanged { quarantine: bool },
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn enroll_armed_defaults_to_none_and_parses() {
        // Onboarding completion arms enforcement.
        let r: Request =
            serde_json::from_str(r#"{"op":"enroll","url":"u","key":"k","armed":true}"#).unwrap();
        match r {
            Request::Enroll { armed, install, .. } => {
                assert_eq!(armed, Some(true));
                assert!(install, "install still defaults true");
            }
            other => panic!("wrong variant: {other:?}"),
        }
        // A base enroll (onboarding not complete) can omit it → None (keep prior).
        let r: Request = serde_json::from_str(r#"{"op":"enroll","url":"u","key":"k"}"#).unwrap();
        match r {
            Request::Enroll { armed, .. } => assert_eq!(armed, None),
            other => panic!("wrong variant: {other:?}"),
        }
    }

    #[test]
    fn disposition_parses_with_and_without_rename() {
        // Client sends a rename on conflict resubmit.
        let r: Request = serde_json::from_str(
            r#"{"op":"disposition","name":"foo","agent":"cursor","choice":"send_to_ew","rename":"foo2"}"#,
        )
        .unwrap();
        match r {
            Request::Disposition {
                name,
                agent,
                choice,
                rename,
                submit_config,
                register,
            } => {
                assert_eq!(name, "foo");
                assert_eq!(agent.as_deref(), Some("cursor"));
                assert!(matches!(choice, Choice::SendToEw));
                assert_eq!(rename.as_deref(), Some("foo2"));
                assert!(submit_config.is_none());
                assert_eq!(register, None, "absent register defers to the role");
            }
            other => panic!("wrong variant: {other:?}"),
        }

        // Older / no-rename clients still parse (serde default).
        let r: Request =
            serde_json::from_str(r#"{"op":"disposition","name":"bar","choice":"skip"}"#).unwrap();
        match r {
            Request::Disposition { rename, choice, .. } => {
                assert!(rename.is_none());
                assert!(matches!(choice, Choice::Skip));
            }
            other => panic!("wrong variant: {other:?}"),
        }

        // A client-supplied submit_config (credential-review override) parses.
        let r: Request = serde_json::from_str(
            r#"{"op":"disposition","name":"foo","choice":"send_to_ew","submit_config":{"Stdio":{"command":"npx","args":["-y","{TOKEN}"],"env":{}}}}"#,
        )
        .unwrap();
        match r {
            Request::Disposition { submit_config, .. } => {
                assert!(submit_config.is_some());
            }
            other => panic!("wrong variant: {other:?}"),
        }
    }
}

File: packages/desktop/src/main/detectord/protocol.ts
// TypeScript mirror of the daemon's wire protocol (detectord/crates/
// mcp_detector_daemon/src/protocol.rs). Newline-delimited JSON over the Unix
// socket: one Reply per Request (FIFO), plus unsolicited Event pushes.

export type Choice = 'send_to_ew' | 'skip'

export type Request =
  | {
      op: 'enroll'
      url: string
      key: string
      mcp_url?: string
      agents?: string[]
      secret?: string
      /** false = detect-only (no sealgate install / hooks). Defaults true. */
      install?: boolean
      /** Arm auto-quarantine. Set true only once onboarding completes. */
      armed?: boolean
    }
  | { op: 'status'; refresh?: boolean }
  | { op: 'list_agents' }
  | { op: 'list_servers' }
  | {
      op: 'disposition'
      name: string
      agent?: string
      choice: Choice
      rename?: string
      /** For SendToEw: submit this (manually redacted) config instead of the
       *  daemon's discovered one, honoring the credential-review overrides. */
      submit_config?: ServerConfig
      /** Register directly (true) or leave pending approval (false). Omit to
       *  let the daemon decide from the user's role. */
      register?: boolean
    }
  /** Install the sealgate entry + session hooks for these agents only. */
  | { op: 'apply_integrations'; agents: string[] }
  /** Remove the sealgate entry for these agents. */
  | { op: 'revert_integrations'; agents: string[] }
  /** An agent's user-scope config file (path + contents), for display. */
  | { op: 'read_config'; agent: string }
  /** Put quarantined servers back: one by name, or all when omitted. */
  | { op: 'restore_quarantined'; name?: string }
  | { op: 'refresh_policy' }
  | { op: 'verify_secret'; key: string }
  | { op: 'reset_secret'; key: string; confirm: boolean }
  | { op: 'unenroll' }

export interface Status {
  user: string
  enrolled: boolean
  org_id?: string | null
  org_name?: string | null
  email?: string | null
  role?: string | null
  quarantine: boolean
  quarantined_count: number
  armed?: boolean
  /**
   * Whether the DAEMON holds macOS Full Disk Access. `null` off macOS,
   * `undefined` from a daemon predating the field.
   *
   * DIAGNOSTIC ONLY - nothing in the app reads it. The daemon does not need the
   * grant: it never watches `$HOME` or the protected folders, so there is
   * nothing for FDA to unlock. Kept on the wire because it is useful in a bug
   * report and only the daemon can answer it for the daemon (TCC grants are
   * per-binary, and the app is a separate binary with its own signature).
   */
  full_disk_access?: boolean | null
}

export interface AgentInfo {
  name: string
  installed: boolean
  /**
   * Hook bindings this agent has and how many are injected, counted by the
   * daemon with the same checks its injector uses. The app reports coverage
   * from these instead of opening the agent's hook file.
   */
  hooks_total?: number
  hooks_installed?: number
  /** URL of the installed sealgate entry, or null when there isn't one. */
  sealgate_url?: string | null
  /** The agent's user-scope config file, for display and `read_config`. */
  config_path?: string | null
  /**
   * Workspace-level hook targets the daemon found for this agent (one
   * `.vscode/tasks.json` per enumerated VSCode workspace) and how many already
   * carry the SealGate task. The daemon counts these because the targets
   * live in the user's project directories - the app deliberately never opens
   * those. Absent from pre-0.6 daemons, hence optional.
   */
  workspace_hooks_total?: number
  workspace_hooks_installed?: number
  /**
   * Whether SealGate can manage this agent at all, or only report that it's
   * there. False for hosts whose MCP servers are Connectors in the vendor's
   * account (ChatGPT) - nothing local to read, write, hook, or proxy.
   *
   * Absent from daemons predating the field, and absence must read as `true`:
   * every agent that existed before it is manageable, and defaulting the other
   * way would silently drop real clients out of setup status.
   */
  manageable?: boolean
}

/** One discovered server instance. `state`: sealgate | known | new | opaque | report. */
// Mirrors the daemon's externally-tagged mcp_detector_lib::ServerConfig.
export type HttpKind = 'Http' | 'Sse' | 'StreamableHttp'
export type OpaqueReason = 'ExtensionProvider' | 'ExtensionServer' | 'CursorPlugin'
export type ServerConfig =
  | { Stdio: { command: string; args: string[]; env: Record<string, string> } }
  | { Http: { url: string; headers: Record<string, string>; kind: HttpKind } }
  | { Opaque: { removable: boolean; reason: OpaqueReason } }

export interface ServerView {
  name: string
  agent: string
  kind: string // stdio | http | opaque
  state: string
  fingerprint?: string | null
  path: string
  config?: ServerConfig | null
}

/** What installing or removing the sealgate entry did for one agent. */
export interface IntegrationChange {
  agent: string
  /** The config file written; null when the agent's own CLI owns the path. */
  path?: string | null
  /** The one-time backup taken before our first edit, if it exists. */
  backup_path?: string | null
  ok: boolean
  error?: string | null
}

export interface SecretOutcome {
  valid?: boolean | null
  expired?: boolean | null
  deleted?: number | null
}

export type Reply =
  | ({ reply: 'status' } & Status)
  | { reply: 'integrations'; changes: IntegrationChange[] }
  | { reply: 'config'; path: string; content: string | null }
  | { reply: 'restored'; restored: number; errors: string[] }
  | { reply: 'agents'; agents: AgentInfo[] }
  | { reply: 'servers'; servers: ServerView[] }
  | ({ reply: 'secret' } & SecretOutcome)
  | { reply: 'ack' }
  | { reply: 'error'; message: string }

export type DetectordEvent =
  | ({ event: 'quarantined' } & ServerView)
  | ({ event: 'discovered' } & ServerView)
  | { event: 'policy_changed'; quarantine: boolean }

/** A line from the daemon is either a Reply or an Event. */
export function isEvent(msg: unknown): msg is DetectordEvent {
  return typeof msg === 'object' && msg !== null && 'event' in msg
}

export function isReply(msg: unknown): msg is Reply {
  return typeof msg === 'object' && msg !== null && 'reply' in msg
}
